blockly-core/core/ui/block_space/block_space_editor.js

```diff
@@ -265,62 +265,55 @@ Blockly.BlockSpaceEditor.prototype.getDeleteAreas = function() {
  * 2. delete blocks on top of flyout
  * @private
  */
-Blockly.BlockSpaceEditor.prototype.bumpOrDeleteOutOfBoundsBlocks_ = function() {
+Blockly.BlockSpaceEditor.prototype.bumpBlocksIntoView_ = function() {
+  // Immediately return if dragging, to avoid expensive calculations
   if (Blockly.Block.isDragging()) {
     return;
   }
 
+  // Can't bump if we don't know our workspace dimensions
   var metrics = this.blockSpace.getMetrics();
-  if (!metrics || metrics.contentWidth > metrics.viewWidth ||
-      metrics.contentHeight > metrics.viewHeight) {
-    // Don't try to bump if there are no metrics, or if the viewWidth/Height is
-    // so small that the content overflows on both sides.
-    return;
-  }
-
-  var oneOrMoreBlocksOutOfBounds = metrics.contentTop < 0 ||
-    metrics.contentTop + metrics.contentHeight >
-    metrics.viewHeight + metrics.viewTop ||
-    metrics.contentLeft < (Blockly.RTL ? metrics.viewLeft : 0) ||
-    metrics.contentLeft + metrics.contentWidth >
-    metrics.viewWidth + (Blockly.RTL ? 2 : 1) * metrics.viewLeft;
-
-  if (!oneOrMoreBlocksOutOfBounds) {
+  if (!metrics) {
     return;
   }
 
+  /** @const */
   var MARGIN = 25;
+
+  /** @const */
   var MARGIN_TOP = 15;
-  var overflow;
-  var blocks = this.blockSpace.getTopBlocks(false);
-  for (var b = 0, block; block = blocks[b]; b++) {
-    var blockXY = block.getRelativeToSurfaceXY();
+
+  // Calculate bounds of view, including bump padding
+  var viewInnerTop = metrics.viewTop + MARGIN_TOP;
+  var viewInnerLeft = metrics.viewLeft + MARGIN;
+  var viewInnerBottom = metrics.viewTop + metrics.viewHeight - MARGIN;
+  var viewInnerRight = metrics.viewLeft + metrics.viewWidth - MARGIN
+  var viewInnerWidth = viewInnerRight - viewInnerLeft;
+  var viewInnerHeight = viewInnerBottom - viewInnerTop;
+
+  // Check every block, and bump if needed.
+  this.blockSpace.getTopBlocks(false).forEach(function (block) {
+    // Skip block if it doesn't fit in the view anyway.
     var blockHW = block.getHeightWidth();
-    // Bump any block that's above the top back inside.
-    overflow = metrics.viewTop + MARGIN_TOP -
-      blockXY.y;
-    if (overflow > 0) {
-      block.moveBy(0, overflow);
+    if (blockHW.width > viewInnerWidth || blockHW.height > viewInnerHeight) {
+      return;
     }
```

Comment:
This check should preserve @joshlory 's original fix - we never bump blocks that are bigger than the workspace itself in either dimension.  Prevents bizarre behavior in the case where this runs before everything is initialized.